src/logging.c

```diff
@@ -134,17 +160,102 @@ void rcutils_logging_console_output_handler(
     message_buffer = (char *)dynamic_buffer;
   }
 
-  if (!location) {
-    fprintf(stream, "[%s] [%s]: %s\n", severity_string, name, message_buffer);
-  } else {
-    fprintf(
-      stream, "[%s] [%s]: %s (%s() at %s:%zu)\n", severity_string, name, message_buffer,
-      location->function_name, location->file_name, location->line_number);
+  // Process the format string looking for known tokens
+  char token_start_delimiter = '{';
+  char token_end_delimiter = '}';
+  char static_output_buffer[1024];
+  memset(static_output_buffer, '\0', sizeof(static_output_buffer));
+  // Start with a fixed size buffer and if during token expansion we need longer,
+  // we'll dynamically allocate space.
+  char * output_buffer = static_output_buffer;
+  size_t output_buffer_size = sizeof(static_output_buffer);
+  const char * str = g_rcutils_logging_output_format_string;
+  size_t size = strlen(g_rcutils_logging_output_format_string);
+
+  int n = 0;
+  for (size_t i = 0; i < size; ++i) {
+    if (str[i] != token_start_delimiter) {
+      // TODO(dhood) before merge: dynamically allocate more space if required.
+      strncat(output_buffer, str + i, 1);
+      continue;
+    }
+    // Found a token start delimiter: determine if there's a token or not.
+    char token[1024];  // No token can be longer than the max format string length.
+    memset(token, '\0', sizeof(token));
+    char static_token_buffer[1024];  // Tokens can't expand to more than this; they'll be truncated.
+    char * token_buffer = static_token_buffer;
+    size_t j;
+    // Look for a token end delimiter.
+    for (j = i + 1; j < size && str[j] != token_end_delimiter; j++) {
+    }
+    if (j >= size) {
+      // No end delimiters found in the remainder of the format string;
+      // there won't be any more tokens so shortcut the rest of the checking
+      // TODO(dhood) before merge: dynamically allocate more space if required.
+      strncat(output_buffer, token_buffer, n);
+      break;
+    }
+    // Found what looks like a token; determine if it's recognized.
+    size_t token_len = j - i - 1;  // not including delimiters
+    strncpy(token, str + i + 1, token_len);
+    if (strcmp("severity", token) == 0) {
+      n = rcutils_snprintf(token_buffer, sizeof(static_token_buffer), "%s", severity_string);
```

Comment:
Using `snprintf` for existing strings isn't necessary. It could simply `memcpy` the characters. It would also be easier to directly copy them to the `output_buffer` rather than using another buffer (`token_buffer`).

Same below - also for the other token types.

Reply:
the reason for the intermediate token buffer instead of directly copying to the output buffer was to avoid the need for checking if dynamic allocation is necessary in multiple places. I'll make a macro for doing that check so we can go straight into the output buffer and remove the token buffer completely.